src/main/java/org/inferred/freebuilder/processor/Analyser.java

```diff
@@ -166,10 +170,88 @@ Metadata analyse(TypeElement type) throws CannotGenerateCodeException {
       metadataBuilder
           .clearProperties()
           .addAllProperties(codeGenerators(properties, baseMetadata, builder.get()));
+
+      metadataBuilder.addAllSuperBuilderTypes(superBuilders(type));
+      metadataBuilder.putAllSuperTypeProperties(
+              processSuperTypeProperties(type, baseMetadata, builder));
     }
     return metadataBuilder.build();
   }
 
+  private ImmutableMap<ParameterizedType, ImmutableList<Property>> processSuperTypeProperties(
+          TypeElement type,
+          Metadata baseMetadata,
+          Optional<TypeElement> builder) throws CannotGenerateCodeException {
+    Map<ParameterizedType, ImmutableList<Property>> toRet =
+            new HashMap<ParameterizedType, ImmutableList<Property>>();
+
+    final ImmutableSet<TypeElement> superTypes = MethodFinder.getSupertypes(type);
+    for (TypeElement superType : superTypes) {
+      if (superType.equals(type)) {
+        continue;
+      }
+
+      final ImmutableSet<ExecutableElement> superMethods = methodsOn(superType, elements);
+      final Map<ExecutableElement, Property> superPropertiesRet =
+              findProperties(superType, superMethods);
+      if (superPropertiesRet.isEmpty()) {
+        continue;
+      }
+
+      ParameterizedType pType = QualifiedName.of(superType).withParameters(
+              superType.getTypeParameters());
+
+      // Code builder dance
+      if (builder.isPresent()) {
+        final Metadata metadataSuperType = analyse(superType);
```

Comment:
We've already figured out what properties are implied by what getter methods; we should just check to see which of those methods are declared abstract on this supertype, rather than regenerating everything.

Also, you can't call analyse on non-FreeBuilder-annotated types, as they shouldn't be expected to follow FreeBuilder's rules about abstract methods.